### PR TITLE
[pytest/hash] Enhance hash test

### PR DESCRIPTION
**Pull request by William-zx**
### Description of PR

Summary:
Made the following change:

1. [Rewrite hash test #1506](https://github.com/Azure/sonic-mgmt/pull/1506)
2. [Complete hash key ingress-port test #1509](https://github.com/Azure/sonic-mgmt/pull/1509)
3. [Generate available send packet ports instead of hard coding src_ports in hash_test.py. #1511](https://github.com/Azure/sonic-mgmt/pull/1511)
4. [Add some hash keys: 'src-mac', 'dst-mac', 'ip-proto', 'vlan-id'. #1512](https://github.com/Azure/sonic-mgmt/pull/1512)

### Type of change

- [] Bug fix
- [] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### How did you verify/test it?
pytest --pdb -s -vvv --capture=no --disable_loganalyzer --inventory veos --host-pattern all -user admin --testbed vms-t1 --testbed_file testbed.csv  --duration=0 --show-capture=stdout  fib/test_fib.py::Test_Hash
============================================================== test session starts ==============================================================
platform linux2 -- Python 2.7.12, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /usr/bin/python
cachedir: .pytest_cache
ansible: 2.0.0.2
rootdir: /var/root/sonic-mgmt/tests, inifile: pytest.ini
plugins: ansible-2.0.2
collected 2 items                                                                                                                               

fib/test_fib.py::Test_Hash::test_hash_ipv4 PASSED
fib/test_fib.py::Test_Hash::test_hash_ipv6 PASSED

============================================================ slowest test durations =============================================================
670.56s call     fib/test_fib.py::Test_Hash::test_hash_ipv6
634.48s call     fib/test_fib.py::Test_Hash::test_hash_ipv4
11.01s setup    fib/test_fib.py::Test_Hash::test_hash_ipv4
2.20s teardown fib/test_fib.py::Test_Hash::test_hash_ipv6
0.60s teardown fib/test_fib.py::Test_Hash::test_hash_ipv4
0.42s setup    fib/test_fib.py::Test_Hash::test_hash_ipv6
========================================================== 2 passed in 1319.29 seconds ==========================================================

### CODE REVIEW

File: ansible/roles/test/files/ptftests/hash_test.py [hunk 1]
```diff
@@ -98,63 +74,23 @@ def check_hash(self, src_ip_range, dst_ip_range, ipv4=True):
...
+        for _ in range(0, self.BALANCING_TEST_TIMES):
+            src_port = random.choice([port for port in self.src_ports if port not in exp_port_list]) if hash_key == "ingress-port" else in_port
+            logging.info("src_port: {}".format(src_port))
+            (matched_index, _) = self.check_ip_route(hash_key, src_port, dst_ip, exp_port_list)
+            hit_count_map[matched_index] = hit_count_map.get(matched_index, 0) + 1
+        logging.info("hit count map: {}".format(hit_count_map))
+
+        # The sample is too little for hash_key vlan-id and ingress-port, check it loose
+        check_balancing_mode = 'loose' if hash_key in ['vlan-id', 'ingress-port'] else 'strict'
+        self.check_balancing(next_hop.get_next_hop(), hit_count_map, check_balancing_mode)
```
Comment (lguohan) [inline ansible/roles/test/files/ptftests/hash_test.py]: i am not sure why we introduce loose mode, it seems defeating the purpose of the test.

**Comment by lguohan** [comment]
@William-zx , there  are many enhancement in the pr. it is very difficult to review all enhancement in one pr. can you break your enhancement in three PRs? We'll review them one by one.

File: ansible/roles/test/files/ptftests/hash_test.py (hunk 1 above)
Comment (William-zx) [reply ansible/roles/test/files/ptftests/hash_test.py]: Load balance is established in the case of a large number of samples. When the sample size is too small, the result may be hashed to the same destination, and the result is load balance not very well. In our hash test, the number of vlan-id and ingress-port was small, so i add them to the loose mode.

**Comment by William-zx** [comment]
> @William-zx , there are many enhancement in the pr. it is very difficult to review all enhancement in one pr. can you break your enhancement in three PRs? We'll review them one by one.

Break the enhancement in four PRs:
1. [Rewrite hash test #1506](https://github.com/Azure/sonic-mgmt/pull/1494)
2. [Complete hash key ingress-port test #1509](https://github.com/Azure/sonic-mgmt/pull/1509)
3. [Generate available send packet ports instead of hard coding src_ports in hash_test.py. #1511](https://github.com/Azure/sonic-mgmt/pull/1511)
4. [Add some hash keys: 'src-mac', 'dst-mac', 'ip-proto', 'vlan-id'. #1512](https://github.com/Azure/sonic-mgmt/pull/1512)

**Comment by William-zx** [comment]
four sub PRs have been merged into the master. close this PR

**Pull request closed without merging**